### DISCUSSION TITLE
Material stream instead of GasMixture and Component instead of Gas

### DISCUSSION THREAD

**Discussion by kookma** [wigging/chemics#33] [Ideas]
Few suggestions and questions:

Why there is a gas and gas mixture? What if chemics use, component and material stream concept? A material stream can accept composition, temperature, pressure, flowrate.
The current implementation uses GasMixture which is like a material stream, but temperature and pressure are set for Gas (actually a component). What does happen if a GasMixture contains gases with different temperatures and pressures?

**Comment by wigging** [comment]
> Why there is a gas and gas mixture? What if chemics use, component and material stream concept? A material stream can accept composition, temperature, pressure, flowrate.

There are two separate class objects for Gas and GasMixture because a different correlation is used for the gas mixture viscosity. Having two separate classes made it easier to separate the gas mixture code from the single gas component code. Regarding your component and material stream idea, do you have any examples of such an approach?

> The current implementation uses GasMixture which is like a material stream, but temperature and pressure are set for Gas (actually a component). What does happen if a GasMixture contains gases with different temperatures and pressures?

The GasMixture class uses the properties of the individual Gas components to calculate the molecular weight of the mixture and the viscosity of the mixture. Each gas component can have different temperatures and pressures. The example below has hydrogen gas at 823 K and nitrogen gas at 773 K. Are you suggesting that the temperature and pressure of the mixture should also be calculated? Or are you suggesting that the user define the temperature and pressure of the mixture?

```python
import chemics as cm
gas1 = cm.Gas('H2', 823)
gas2 = cm.Gas('N2', 773)
gas_mixture = cm.GasMixture([gas1, gas2], [0.85, 0.15])
```

**Reply by kookma** [reply]
> Are you suggesting that the temperature and pressure of the mixture should also be calculated? Or are you suggesting that the user define the temperature and pressure of the mixture?
> 

Well yes! a gas mixture is created from two gas at different temperature and pressure! at least energy balance is needed here! If you define these as stream it make sense to calculate the mixture temperature but as a single compound no.

**Reply by wigging** [reply]
The `GasMixture` class uses correlations that calculate the mixture viscosity based on the individual gas components. See the [chemics docs](https://chemics.readthedocs.io/en/latest/api/gas_mixture.html) for more information. The reference where these correlations were obtained is not clear on how the individual gas component properties should be determined. But it looks like the current implementation in chemics can be improved.

**Comment by kookma** [comment]
> Regarding your component and material stream idea, do you have any examples of such an approach?

Yes, almost all process simulator use this concept for example Aspen plus, Aspen Hysys, Pro/II, Symmetry, DWSIM, etc...

For example in [DWSIM](https://github.com/DanWBR/dwsim) which is an open source process simulator see: https://www.youtube.com/watch?v=MSo76MJUdlQ

```
'Create a new Material Stream object
Dim ms As DWSIM.SimulationObjects.Streams.MaterialStream
ms = My.Application.ActiveSimulation.Collections.CLCS_MaterialStreamCollection.Add()
ms.GraphicObject.Tag = "Feed"
ms.GraphicObject.X = 30
ms.GraphicObject.Y = 30

'Set the stream pressure, temperature, mass flow and composition
ms.Phases(0).Properties.pressure = 101325 'Pa
ms.Phases(0).Properties.temperature = 298.15 'K
ms.Phases(0).Properties.massflow = 1 'kg/s
ms.Phases(0).Properties.molarfraction = 1
ms.Phases(2).Properties.molarfraction = 0.21 'O2
ms.Phases(2).Properties.molarfraction = 0.79 'N2
````

or in Aspen [Hysys](https://www.chemengonline.com/using-excel-vba-process-simulator-data-extraction/)

```
'Create a reference to the Hysys application
Dim hyApp As Hysys.Application
Set hyApp = CreateObject("Hysys.Application")

'Create a reference to the active simulation case
Dim ProjectSimulation As SimulationCase
Set ProjectSimulation = hyApp.ActiveDocument

'Create a reference to the material stream object named "100"
Dim HMB_stream As MaterialStream
Set HMB_stream = ProjectSimulation.Flowsheet.MaterialStreams.Item("100")

'Extract the pressure, mass flowrate and component mass fractions from the stream
Range("B2").Value = HMB_stream.Pressure.GetValue("kPa")
Range("B3").Value = HMB_stream.MassFlow.GetValue("kg/h")
Range("B4").Value = HMB_stream.ComponentMassFraction.GetValue("Water", "")
Range("B5").Value = HMB_stream.ComponentMassFraction.GetValue("Ethanol", "")
Range("B6").Value = HMB_stream.ComponentMassFraction.GetValue("Acetone", "")
```

**Reply by wigging** [reply]
I see how a material stream approach would be useful for an engineer or scientist. But the core components of chemics need to be further developed before such a feature can be added to the package.

**Reply by kookma** [reply]
I will keep an eye on this and would be happy to test or help when I can.

**Comment by kookma** [comment]
It seems coolprop is the closest to what I said. See: https://tahircpe.github.io/Material_Stream_Class/
